tree/treeplayer/inc/ROOT/TVec.hxx

```diff
@@ -0,0 +1,381 @@
+}
+
+template <typename T, typename V>
+TVec<char> operator<(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "<", [](const T& t, const V &v) -> char {return t<v;});
+}
+
+// Workaround for operators >&co on vectors
+template <typename T>
+TVec<char> operator>(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator><T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator>=(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator>=<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator==(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator==<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator!=(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator!=<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator<=(const TVec<T> &v0, const TVec<T> &v1)
+{
+   return ROOT::Experimental::VecOps::operator<=<T, T>(v0, v1);
+}
+template <typename T>
+TVec<char> operator<(const TVec<T> &v0, const TVec<T> &v1)
+{
+   // clang-format off
+   return ROOT::Experimental::VecOps::operator< <T, T>(v0, v1);
+   // clang-format on
+}
+
+///@}
+
+/** @name Math Functions
+ *  Math functions on TVecs
+*/
+///@{
+#define MATH_FUNC(FUNCNAME)\
+template <typename T>\
+auto FUNCNAME(const TVec<T> &v) -> TVec<decltype(std::FUNCNAME(v[0]))>\
+{return ROOT::Internal::VecOps::Operate(v, [](const T& t) {return std::FUNCNAME(t);});}
+
+MATH_FUNC(sqrt)
+MATH_FUNC(log)
+MATH_FUNC(sin)
+MATH_FUNC(cos)
+MATH_FUNC(asin)
+MATH_FUNC(acos)
+MATH_FUNC(tan)
+MATH_FUNC(atan)
+MATH_FUNC(sinh)
+MATH_FUNC(cosh)
+MATH_FUNC(asinh)
+MATH_FUNC(acosh)
+MATH_FUNC(tanh)
+MATH_FUNC(atanh)
+
+
+///@}
+
+/// Inner product
+template <typename T, typename V>
+auto Dot(const TVec<T> v0, const TVec<V> v1) -> decltype(v0[0] * v1[0])
+{
+   ROOT::Internal::VecOps::CheckSizes(v0.size(), v1.size(), "Dot");
+   return std::inner_product(v0.begin(), v0.end(), v1.begin(), decltype(v0[0] * v1[0])(0));
+}
+
+/// Sum elements
+template <typename T>
+T Sum(const TVec<T> v)
+{
+   return std::accumulate(v.begin(), v.end(), 0);
+}
+
+template <class T>
+std::ostream &operator<<(std::ostream &os, const TVec<T> &v)
+{
+   // In order to print properly, convert to 64 bit int if this is a char
+   using Print_t = typename std::conditional<ROOT::Internal::VecOps::TIsChar<T>::value, long long int, T>::type;
+   os << "{ ";
+   auto size = v.size();
+   if (size) {
+      for (size_t i = 0; i < size - 1; ++i) {
+         os << (Print_t)v[i] << ", ";
+      }
+      os << (Print_t)v[size - 1];
+   }
+   os << " }";
+   return os;
+}
+
+template <typename T>
+TVec<T> Filter(const TVec<T> &v, bool cond)
+{
+   return cond ? v : TVec<T>();
+}
+
+template <typename T, typename V>
+TVec<T> Filter(const TVec<T> &v, const TVec<V> &conds)
```

Comment:
This is also in #1612 :)

Reply:
I am wondering if this is bad or good. Probably the two can coexist? in #1612 we have the signature which accommodates a callable which is more general.
I am not sure yet what is the right way to proceed here...

Reply:
I'm for moving `Map` and `Filter` from #1612 over here -- or whatever you prefer! :)

Reply:
`std::copy` and `std::copy_if` are c++11, the problem is that they have horrible signatures. imho `Filter` should wrap `copy_if` just as `Operate` here maps `std::transform` (see #1612 ), but I think @dpiparo is going to merge the two PRs or something like that

Reply:
The issue here is that copy_if does not allow to have, without hacks, iteration on 2 collections but rather 1 collection and a predicate.
Here the conditions according to which individual elements of the TVec have to be retained are packed in a TVec too.

Reply:
But packing in a `TVec` means that you have to loop once to compute the `conds`, and then a second time to copy things. If you use a lambda as predicate, then you can probably get away with a single loop, no?

Reply:
yes, as things stand, `v[v>3]` loops twice, `Filter(v, [](int v) { return v > 3; })` loops once. both options will be available if I understand correctly, with obvious tradeoffs